polar-core/src/data_filtering.rs

```diff
@@ -63,737 +67,635 @@ pub struct FetchRequest {
+        }
 
-            match (lhs.value(), rhs.value()) {
-                // Unifying two variables
-                (Value::Variable(l), Value::Variable(r)) => {
-                    var_info.cycles.push((l.clone(), r.clone()));
-                }
-                // Unifying a variable with a value
-                (Value::Variable(var), val) | (val, Value::Variable(var)) => var_info
-                    .eq_values
-                    .push((var.clone(), Term::new_temporary(val.clone()))),
-                // Unifying something else.
-                // 1 = 1 is irrelevant for data filtering, other stuff seems like an error.
-                // @NOTE(steve): Going with the same not yet supported message but if this is
-                // coming through it's probably a bug in the simplifier.
-                (_a, _b) => {
-                    unimplemented!(
-                        "Unification of values is not yet supported for data filtering."
-                    );
-                }
-            };
+        /// get the id for this variable, or create one if the variable is new.
+        fn get_var_id(
+            vars: &mut HashMap<Id, HashSet<Symbol>>,
+            var: Symbol,
+            counter: &Counter,
+        ) -> Id {
+            seek_var_id(vars, &var).unwrap_or_else(|| {
+                let new_id = counter.next();
+                let mut new_set = HashSet::new();
+                new_set.insert(var);
+                vars.insert(new_id, new_set);
+                new_id
+            })
         }
-        Operator::In => {
-            assert_eq!(exp.args.len(), 2);
 
-            let mut lhs = exp.args[0].clone();
-            if let Value::Expression(op) = lhs.value() {
-                lhs = process_exp(var_info, op).unwrap();
-            };
+        let counter = info.counter;
+        let mut cycles = info.cycles;
+        let fields = info.field_relationships;
 
-            let mut rhs = exp.args[1].clone();
-            if let Value::Expression(op) = rhs.value() {
-                rhs = process_exp(var_info, op).unwrap();
-            };
+        // if x.y=a and x.y=b then a=b, so add a cycle.
+        fields
+            .iter()
+            .flat_map(|(p1, f1, c1)| {
+                fields.iter().filter_map(move |(p2, f2, c2)| {
+                    (c1 != c2 && f1 == f2 && p1 == p2).then(|| (c1.clone(), c2.clone()))
+                })
+            })
+            .for_each(|p| cycles.push(p));
+
+        // group the variables into equivalence classes.
+        let mut variables = partition_equivs(cycles)
+            // Give each cycle an id
+            .into_iter()
+            .map(|c| (counter.next(), c))
+            .collect::<HashMap<_, _>>();
+
+        let mut assign_id = |item| get_var_id(&mut variables, item, &counter);
+
+        // now convert the remaining VarInfo fields into equivalent Vars fields.
+
+        let in_relationships = info
```

Comment:
The reason this was down lower before is because we don't know if `p1 = p2` yet. They could be `foo` and `bar` and maybe somewhere else there's a `foo = bar` but if we do this check before knowing that we'll miss it.

Reply:
We should probably add a test for that case since it's only in my mind.

Reply:
are you thinking of `field_relationships`?